subsys/caf/modules/sensor_sampler.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <drivers/sensor.h>
+
+#include "sensor_event.h"
+
+#include CONFIG_CAF_SENSOR_SAMPLER_DEF_PATH
+
+#define MODULE sensor_sampler
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_CAF_SENSOR_SAMPLER_LOG_LEVEL);
+
+#define SAMPLE_THREAD_STACK_SIZE	CONFIG_CAF_SENSOR_SAMPLER_THREAD_STACK_SIZE
+#define SAMPLE_THREAD_PRIORITY		CONFIG_CAF_SENSOR_SAMPLER_THREAD_PRIORITY
+
+struct sensor_data {
+	const struct device *dev;
+	int sampling_period;
+	int remaining_wait;
+};
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_ERROR
+};
+
+static struct sensor_data sensor_data[ARRAY_SIZE(sensor_configs)];
+
+static enum state state;
+
+static K_THREAD_STACK_DEFINE(sample_thread_stack, SAMPLE_THREAD_STACK_SIZE);
+static struct k_thread sample_thread;
+static int64_t uptime;
+
+
+static void report_error(void)
+{
+	state = STATE_ERROR;
+	module_set_state(MODULE_STATE_ERROR);
+}
+
+static void send_sensor_event(const char *descr, struct sensor_value *data, size_t data_cnt)
+{
+	struct sensor_event *event = new_sensor_event(sizeof(float) * data_cnt);
+	float *data_ptr = sensor_event_get_data_ptr(event);
+
+	event->descr = descr;
+
+	__ASSERT_NO_MSG(sensor_event_get_data_cnt(event) == data_cnt);
+	for (size_t i = 0; i < data_cnt; i++) {
+		data_ptr[i] = sensor_value_to_double(&data[i]);
+	}
+
+	EVENT_SUBMIT(event);
+}
+
+static int sample_sensor(const struct device *dev, const struct sensor_config *sc)
+{
+	size_t data_cnt = 0;
+
+	for (size_t i = 0; i < sc->chan_cnt; i++) {
+		data_cnt += sc->chans[i].data_cnt;
+	}
+
+	int err = 0;
+	size_t data_idx = 0;
+	struct sensor_value data[data_cnt];
```

Comment:
why not storing the data directly in the sensor event?

Reply:
The sensor event uses floats to represent the data. Zephyr sensor API uses pointer to buffer of `struct sensor_value`, so we cannot write the data directly to sensor event.

Reply:
I could make the buffer smaller (as big as data count for single channel) and combine sampling sensor and filling sensor data into one function. Still, we will need to use the buffer and functionalities will no longer be separated.

Reply:
Yes, I meant if maybe we could convert on the fly but I have not very strong opinion on it.

Reply:
Ok, I will change it